exla/lib/exla/backend.ex

```diff
@@ -0,0 +1,268 @@
+defmodule EXLA.Backend do
+  @moduledoc """
+  A Nx tensor backend for the data kept on the device.
+
+  You can directly transfer to this backend by calling
+  `Nx.backend_transfer/2` or `Nx.backend_copy/2`. It
+  allows the following options:
+
+    * `:client` - the client to store the data on.
+      Defaults to the client configured in `Nx.Defn`,
+      otherwise uses `:host`.
+
+    * `:device_id` - which device to store it on
+
+  To get the data out of the device backend into a regular
+  tensor, call `Nx.backend_transfer/1` (with the device
+  tensor as the single argument).
+  """
+
+  @behaviour Nx.Backend
+  @enforce_keys [:buffer]
+  defstruct [:buffer]
+
+  alias Nx.Tensor, as: T
+  alias EXLA.Backend, as: B
+
+  import Nx.Shared
+
+  @impl true
+  def constant(%{type: type, shape: shape} = out, constant, backend_options) do
+    data = :binary.copy(number_to_binary(constant, type), Nx.size(shape))
+    from_binary(out, data, backend_options)
+  end
+
+  defp number_to_binary(number, type),
+    do: match_types([type], do: <<write!(number, 0)>>)
+
+  @impl true
+  def from_binary(%T{shape: shape, type: type} = tensor, binary, opts) do
+    {client, device_id} = client_and_device_id(opts)
+    shape = EXLA.Shape.make_shape(type, shape)
+    buffer = EXLA.DeviceBuffer.place_on_device(binary, shape, client, device_id)
+    put_in(tensor.data, %B{buffer: buffer})
+  end
+
+  @impl true
+  def backend_copy(tensor, Nx.Tensor, opts) do
+    backend_copy(tensor, Nx.BinaryBackend, opts)
+  end
+
+  # TODO: Support direct transfers without going through Elixir
+  def backend_copy(%T{data: %B{buffer: buffer}} = tensor, backend, opts) do
+    backend.from_binary(tensor, EXLA.DeviceBuffer.read(buffer), opts)
+  end
+
+  @impl true
+  def backend_transfer(%T{data: %B{buffer: buffer}} = tensor, backend, opts) do
+    if backend == __MODULE__ and same_client_device?(buffer, opts) do
+      tensor
+    else
+      try do
+        backend_copy(tensor, backend, opts)
+      after
+        EXLA.DeviceBuffer.deallocate(buffer)
+      end
+    end
+  end
+
+  @impl true
+  def backend_deallocate(%T{data: %B{buffer: buffer}}) do
+    EXLA.DeviceBuffer.deallocate(buffer)
+  end
+
+  @impl true
+  def to_batched_list(out, tensor, opts) do
+    binary_tensor = backend_copy(tensor, Nx.BinaryBackend, [])
+    Nx.BinaryBackend.to_batched_list(out, binary_tensor, opts)
+  end
```

Comment:
Is there any issue with this approach?

Reply:
I think so. I don't think we want to convert to a list of binarybackend tensors. If possible we would still keep them on the device... but I am not sure how easy it would be to implement it. So my suggestion is to raise on this function for now? 🤔

Reply:
My initial thought was that we could generate an expr that returns a tuple of slices, but with this approach the computation graph would grow with data size or batch granularity, right?